File: bigquery/BQSample.go
/*
Package main demonstrates various Bigquery API operations.

To use, specify the project and a corresponding JSON service account file
with permissions on the target project.  If running from a
Google Compute Engine (GCE) VM, uncomment the corresponding section below
and ensure the GCE instance is created with a scope that is authorized for
the logging API.  For Google App Engine (GAE), add in the additional
imports and initialize the context using the request handler.

https://godoc.org/golang.org/x/oauth2/google
*/
package main

import (
	//"fmt"
	"log"
	"time"
	//"net/http"     // import for GAE and GCE

	// import for GAE
	/*
	   "google.golang.org/appengine"
	   "google.golang.org/appengine/urlfetch"
	*/

	"golang.org/x/net/context"
	"golang.org/x/oauth2"
	"golang.org/x/oauth2/google"
	"google.golang.org/api/bigquery/v2"
)

const (
	projectId              = "YOUR_PROJECT_ID"
	serviceAccountJSONFile = "YOUR_SERVICE_ACCOUNT_JSON_FILE"
)

func listDatasets(service *bigquery.Service, projectId string) {
	nextPageToken := ""
	for {
		resp, err := service.Datasets.List(projectId).PageToken(nextPageToken).Do()
		if err != nil {
			log.Fatalf("Unable to list Datasets: %s", err)
		}
		for _, ds := range resp.Datasets {
			log.Printf("Datasets %v", ds.Id)
		}
		nextPageToken = resp.NextPageToken
		if nextPageToken == "" {
			break
		}
	}
}

func startQuery(service *bigquery.Service, projectId string, querySQL string) (jobId *bigquery.JobReference, err error) {
	log.Printf("Inserting Query Job: %v", querySQL)
	bqjob := bigquery.Job{
		Configuration: &bigquery.JobConfiguration{
			Query: &bigquery.JobConfigurationQuery{
				Query: querySQL,
			},
		},
	}
	ins, err := service.Jobs.Insert(projectId, &bqjob).Do()
	if err != nil {
		log.Printf("Unable to acquire token source: %v", err)
		return nil, err
	}
	return ins.JobReference, nil
}

func checkQueryResults(service *bigquery.Service, projectId string, jobId *bigquery.JobReference) (job *bigquery.Job, err error) {
	startTime := time.Now().UnixNano() / 1000000
	var elapsedTime int64 = 0
	for {
		var pollJob, err = service.Jobs.Get(projectId, jobId.JobId).Do()
		if err != nil {
			log.Printf("Unable to acquire jobID: %v", err)
			return nil, err
		}
		elapsedTime = (time.Now().UnixNano() / 1000000) - startTime
		log.Printf("Job status (%vms) %v: %v\n", elapsedTime,
			jobId.JobId, pollJob.Status.State)
		if pollJob.Status.State == "DONE" {
			return pollJob, nil
		}
		// Pause execution for one second before polling job status again, to
		// reduce unnecessary calls to the BigQUery API and lower overall
		// application bandwidth.
		time.Sleep(1000 * time.Microsecond)
	}
}

func displayQueryResults(service *bigquery.Service, projectId string, completedJob *bigquery.Job) {
	queryResult, err := service.Jobs.GetQueryResults(projectId, completedJob.JobReference.JobId).Do()
	if err != nil {
		log.Fatalf("Unable to render query results: %v", err)
	}
	log.Printf("Query Results:------------")
	for _, row := range queryResult.Rows {
		for _, field := range row.F {
			log.Printf("%v", field.V)
		}
	}
}

func Main() {

	// Comment out setting the environment variable while running on GCE or GAE or if using
	// credentials from gcloud
	//os.Setenv("GOOGLE_APPLICATION_CREDENTIALS", serviceAccountJSONFile)
	// End comment

	src, err := google.DefaultTokenSource(oauth2.NoContext, bigquery.BigqueryScope)
	if err != nil {
		log.Fatalf("Unable to acquire token source: %v", err)
	}
	tok, err := src.Token()
	if err != nil {
		log.Fatalf("Credentials Token file can't be read")
	}

	// Use when running in GAE
	// r is the current *http.Request from the handler
	// https://cloud.google.com/appengine/docs/go/requests#Go_Requests_and_HTTP
	/*
	   ctx := appengine.NewContext(r)
	   src, err := google.DefaultTokenSource(ctx, bigquery.BigqueryScope)
	   if err != nil {
	           log.Infof(ctx, "Unable to acquire token source: %v", err)
	   }
	   client := &http.Client{
	           Transport: &oauth2.Transport{
	                   Source: src),
	                   Base:   &urlfetch.Transport{ Context: ctx },
	           },
	   }
	*/

	// For user interactive webflow
	/*
	   conf := &oauth2.Config{
	           ClientID:     "YOUR_CLIENT_ID",
	           ClientSecret: "YOUR_CLIENT_SECRET",
	           RedirectURL:  "urn:ietf:wg:oauth:2.0:oob",
	           Scopes: []string{
	                   bigquery.BigqueryScope,
	           },
	           Endpoint: google.Endpoint,
	   }

	   url := conf.AuthCodeURL("state")
	   log.Printf("Visit the URL for the auth dialog and: %v", url)

	   var code string
	   fmt.Println("Enter code: ")
	   if _, err = fmt.Scan(&code); err != nil {
	           log.Fatalf(err.Error())
	   }
	   tok, err = conf.Exchange(context.Background(), code)
	   if err != nil {
	           log.Fatalf(err.Error())
	   }
	*/
	src = oauth2.ReuseTokenSource(tok, src)

	client := oauth2.NewClient(context.Background(), src)
	service, err := bigquery.New(client)
	if err != nil {
		log.Fatalf("Unable to create logging client: %v", err)
	}

	// First list out all the available datasets
	listDatasets(service, projectId)

	// Start a Query Job
	querySql := "SELECT TOP(word, 50), COUNT(*) FROM publicdata:samples.shakespeare"
	jobId, err := startQuery(service, projectId, querySql)
	if err != nil {
		log.Fatalf("Unable to get submit Query: %v", err)
	}
	completedJob, err := checkQueryResults(service, projectId, jobId)
	if err != nil {
		log.Fatalf("Unable to get submit Query: %v", err)
	}
	log.Printf(completedJob.JobReference.JobId)

	// Return and display the results of the Query Job
	displayQueryResults(service, projectId, completedJob)

}

File: bigquery/BQSample.cs
/*
 * Copyright (c) 2015 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */

/**
 * Example of authorizing with BigQuery and reading from a public dataset
 * using the Google Client Libraries.
 * 
 * Usage:
 * Add projects references using NuGet
 * "Google.Apis.Bigquery.v2"   (other dependencies will get added automatically)
 * Generate and download service account .p12 from place it to c:\
 * specify the projectId, serviceAccountEmail and certificateFile below.
 */

using System;
using System.Collections.Generic;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Security.Cryptography.X509Certificates;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

using Google.Apis;
using Google.Apis.Auth.OAuth2;
using Google.Apis.Bigquery.v2;
using Google.Apis.Bigquery.v2.Data;
using Google.Apis.Services;

namespace BQSample
{
    internal class Program
    {
        [STAThread]
        static void Main(string[] args)
        {
            try
            {
                new Program().Run().Wait();
            }
            catch (AggregateException ex)
            {
                foreach (var err in ex.InnerExceptions)
                {
                    Console.WriteLine("ERROR: " + err.Message);
                }
            }
            Console.ReadKey();
        }

        #region Consts

        private const string projectId = "YOUR_PROJECT_ID";  
        #endregion

        private async Task Run()
        {
            //Authentication Options:  1. JSON ServiceAccountCredential file, 2. ComputeCredential, 3. UserCredential (gcloud application credentials)

            //Set environment variable for 1.JSON ServiceAccountsCredential
            //string CREDENTIAL_FILE = "C:\\YOUR_CRED_FILE.json";

            // 2. ComputeCredentials will get used automatically use while running on GCE 

            //Set environment variable for 3. UserCredentials acquired using gcloud.
            //by default the gcloud credential file can be found at  https://cloud.google.com/sdk/gcloud/#gcloud.auth 
            //to override it, set
            //string CREDENTIAL_FILE = "c:\\application_default_credentials.json"
            //If GOOGLE_APPLICATION_CREDENTIALS is not set, GoogleCredential will attempt to automaticallyg find the gcloud credentials.

            //Environment.SetEnvironmentVariable("GOOGLE_APPLICATION_CREDENTIALS", CREDENTIAL_FILE);
            GoogleCredential credential = await GoogleCredential.GetApplicationDefaultAsync();
            if (credential.IsCreateScopedRequired)
                credential = credential.CreateScoped(new string[] {BigqueryService.Scope.Bigquery});
            
            //Authentication Option 4: non-gcloud User interactive webflow.
            /*
            UserCredential credential;
            string CLIENTSECRETS_LOCATION = "c:\\client_secrets.json";
            using (var stream = new FileStream(CLIENTSECRETS_LOCATION, FileMode.Open, FileAccess.Read))
            {
                credential = await GoogleWebAuthorizationBroker.AuthorizeAsync(
                    GoogleClientSecrets.Load(stream).Secrets,
                    new[] { BigqueryService.Scope.Bigquery }, Environment.UserName, CancellationToken.None);
            }
            */
            //string clientId = "YOUR_CLIENT_ID.apps.googleusercontent.com";
            //string clientSecret = "YOUR_CLIENT_SECRET";
            //credential = await GoogleWebAuthorizationBroker.AuthorizeAsync(new ClientSecrets { ClientId = clientId, ClientSecret = clientSecret },
            //    new[] { BigqueryService.Scope.Bigquery }, Environment.UserName, CancellationToken.None);
            //Console.WriteLine("Credential file saved at: " + Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData));            
            
            var service = new BigqueryService(new BaseClientService.Initializer()
            {
                HttpClientInitializer = credential,
                ApplicationName = "BigQuery Sample",
            });

            Console.WriteLine("Listing Available Datasets on " + projectId);
            listDatasets(service,projectId);
            Console.WriteLine("=============================");

            Console.WriteLine("Running sample query on publicdata");
            string querySql = "SELECT TOP(word, 50), COUNT(*) FROM publicdata:samples.shakespeare";
            Console.WriteLine(querySql);
            JobReference jobId = startQuery(service, projectId, querySql);
            Console.WriteLine("=============================");

            // Poll for Query Results, return result output
            Job completedJob = checkQueryResults(service, projectId, jobId);

            // Return and display the results of the Query Job
            displayQueryResults(service, projectId, completedJob);

        }

        // Display all BigQuery datasets associated with a project.
        private void listDatasets(BigqueryService bigquery, string projectId)
        {
            DatasetList datasetRequest = bigquery.Datasets.List(projectId).Execute();
            string nextPageToken = datasetRequest.NextPageToken;
            do
            { 
                if (datasetRequest.Datasets != null)
                    foreach (var dataset in datasetRequest.Datasets)
                    {
                         Console.WriteLine(dataset.Id);
                    }
            } while (!String.IsNullOrEmpty(nextPageToken));
        }

        // Creates a Query Job for a particular query on a dataset.
        private JobReference startQuery(BigqueryService bigquery, string projectId, string querySql)
        {
            Console.WriteLine("Inserting Query Job: " + querySql);            
            JobConfigurationQuery queryConfig = new JobConfigurationQuery{ Query = querySql };
            JobConfiguration config = new JobConfiguration{ Query = queryConfig };
            Job job = new Job { Configuration = config };
            JobsResource.InsertRequest insert = bigquery.Jobs.Insert(job, projectId);
            JobReference jobId = insert.Execute().JobReference;
            Console.WriteLine("Job ID of Query Job is: " + jobId.JobId);
            return jobId;
        }

      // Polls the status of a BigQuery job, returns Job reference if "Done".
      private Job checkQueryResults(BigqueryService bigquery, String projectId, JobReference jobId) {
        // Variables to keep track of total query time
        double startTime = DateTime.Now.Millisecond;
        while (true) {
          Job pollJob = bigquery.Jobs.Get(projectId, jobId.JobId).Execute();
          var duration = DateTime.Now.Millisecond - startTime;
          Console.WriteLine("Job status: ElapsedTime [" +  duration + "ms]  JobID: [" +
              jobId.JobId + "]  JobState: [" + pollJob.Status.State + "]");
          if (pollJob.Status.State.Equals("DONE")) {
            return pollJob;
          }
          // Pause execution for one second before polling job status again, to
          // reduce unnecessary calls to the BigQUery API and lower overall
          // application bandwidth.
          Thread.Sleep(1000);
        }
      }

      private void displayQueryResults(BigqueryService bigquery, String projectId, Job completedJob) 
      {
        GetQueryResultsResponse queryResult = bigquery.Jobs.GetQueryResults(projectId, completedJob.JobReference.JobId).Execute();
        Console.WriteLine("Query Results:");
        foreach (var row in queryResult.Rows) {
          foreach (var field in row.F) {
              Console.WriteLine(field.V);
           }
          Console.WriteLine();
        }
      }
    }
}
